downloaders/apkmirror-downloader/src/main/kotlin/app/revanced/manager/plugin/downloader/apkmirror/APKMirrorDownloader.kt

```diff
@@ -0,0 +1,106 @@
+import java.net.URI
+import java.nio.file.Files
+import java.util.UUID
+import java.util.zip.ZipFile
+import kotlin.io.path.ExperimentalPathApi
+import kotlin.io.path.deleteRecursively
+import kotlin.io.path.outputStream
+
+@Parcelize
+class ApkMirrorApp(
+    val downloadUrl: DownloadUrl
+) : Parcelable
+
+private object ArscLogger : APKLogger {
+    const val TAG = "ARSCLib"
+
+    override fun logMessage(msg: String) {
+        Log.i(TAG, msg)
+    }
+
+    override fun logError(msg: String, tr: Throwable?) {
+        Log.e(TAG, msg, tr)
+    }
+
+    override fun logVerbose(msg: String) {
+        Log.v(TAG, msg)
+    }
+}
+
+@OptIn(ExperimentalPathApi::class)
+val ApkMirrorDownloader = Downloader<ApkMirrorApp> {
+    get { packageName, version ->
+        ApkMirrorApp(
+            downloadUrl =  runWebView("APKMirror") {
+                download { url, _, userAgent ->
+                    finish(
+                        DownloadUrl(
+                            url,
+                            mapOf("User-Agent" to userAgent)
+                        )
+                    )
+                }
+
+                Uri.Builder()
+                    .scheme("https")
+                    .authority("www.apkmirror.com")
+                    .appendQueryParameter("post_type", "app_release")
+                    .appendQueryParameter("searchtype", "apk")
+                    .appendQueryParameter("s", version?.let { "$packageName $it" } ?: packageName)
+                    .appendQueryParameter("bundles%5B%5D" /* bundles[] */, "apk_files")
+                    .toString()
+            }
+        ) to version
+    }
+
+    download { app, outputStream ->
+        val workingDir = Files.createTempDirectory("apkmirror_dl")
+        try {
+            if (URI(app.downloadUrl.url).path.substringAfterLast('/').endsWith(".apk")) {
+                val (inputStream, size) = app.downloadUrl.toDownloadResult()
+                inputStream.use {
+                    if (size != null) reportSize(size)
+                    it.copyTo(outputStream)
+                }
+            } else {
+                val downloadedFile = workingDir.resolve(UUID.randomUUID().toString()).also {
+                    it.outputStream().use { output ->
+                        app.downloadUrl.toDownloadResult().first.copyTo(output)
+                    }
+                }
+                val xapkWorkingDir = workingDir.resolve("xapk").also { it.toFile().mkdirs() }
+
+                ZipFile(downloadedFile.toString()).use { zip ->
+                    zip.entries().asSequence().forEach { entry ->
+                        xapkWorkingDir.resolve(entry.name).also { it.parent.toFile().mkdirs() }.also { outputFile ->
+                            zip.getInputStream(entry).use { input ->
+                                File(outputFile.toString()).outputStream().use { output ->
+                                    input.copyTo(output)
+                                }
```

Comment:
You could probably use `Files.copy` from Java `nio` to write the file instead of doing this manually